ding/reward_model/trex_reward_model.py

```diff
@@ -0,0 +1,430 @@
+        rand_length = np.random.randint(min_snippet_length, max_snippet_length, size=num_snippets)
+        for n in range(num_snippets):
+            ti = 0
+            tj = 0
+            #only add trajectories that are different returns
+            while (ti == tj):
+                #pick two random demonstrations
+                ti = np.random.randint(num_demos)
+                tj = np.random.randint(num_demos)
+            #create random snippets
+            #find min length of both demos to ensure we can pick a demo no earlier
+            #than that chosen in worse preferred demo
+            min_length = min(len(demonstrations[ti]), len(demonstrations[tj]))
+            if ti < tj:  # pick tj snippet to be later than ti
+                ti_start = np.random.randint(min_length - rand_length[n] + 1)
+                # print(ti_start, len(demonstrations[tj]))
+                tj_start = np.random.randint(ti_start, len(demonstrations[tj]) - rand_length[n] + 1)
+            else:  # ti is better so pick later snippet in ti
+                tj_start = np.random.randint(min_length - rand_length[n] + 1)
+                # print(tj_start, len(demonstrations[ti]))
+                ti_start = np.random.randint(tj_start, len(demonstrations[ti]) - rand_length[n] + 1)
+            traj_i = demonstrations[ti][ti_start:ti_start + rand_length[n]:2
+                                        ]  # skip everyother framestack to reduce size
+            traj_j = demonstrations[tj][tj_start:tj_start + rand_length[n]:2]
+
+            max_traj_length = max(max_traj_length, len(traj_i), len(traj_j))
+            label = int(ti <= tj)
+            self.training_obs.append((traj_i, traj_j))
+            self.training_labels.append(label)
+
+        self._logger.info(("maximum traj length: {}".format(max_traj_length)))
+        return self.training_obs, self.training_labels
+
+    def train(self):
+        # check if gpu available
+        device = self.device  # torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+        # Assume that we are on a CUDA machine, then this should print a CUDA device:
+        self._logger.info("device: {}".format(device))
+        training_inputs, training_outputs = self.training_obs, self.training_labels
+        loss_criterion = nn.CrossEntropyLoss()
+
+        cum_loss = 0.0
+        training_data = list(zip(training_inputs, training_outputs))
+        for epoch in range(self.cfg.reward_model.update_per_collect):  # todo
+            np.random.shuffle(training_data)
+            training_obs, training_labels = zip(*training_data)
+            for i in range(len(training_labels)):
+
+                # traj_i, traj_j has the same length, however, they change as i increases
+                traj_i, traj_j = training_obs[i]  # traj_i is a list of array generated by env.step
+                labels = np.array([training_labels[i]])  # training_labels[i] is a boolean integer: 0 or 1
```

Comment:
redundant np.array

Reply:
Line 317 has been deleted and line 322 has changed into labels = torch.tensor([training_labels[i]]).to(device)